### PR TITLE
GH-43209: [C++] Add lint for DCHECK in public headers

### PR DESCRIPTION
**Pull request by zanmato1984**
### Rationale for this change

I raised my question in #43209 about which I have always been curious. The top answer makes a good sense and after some searching in the code base, I found more evidence like:
https://github.com/apache/arrow/blob/03726178494c8978bf48b9bab15ed9676e7c9196/cpp/src/arrow/public_api_test.cc#L67-L71

So I'm making the following changes to `DCHECK` macro family.

### What changes are included in this PR?

1. Add lint rule for `DCHECK` usage in public headers;
2. Cleanup exisiting `DCHECK` usages in public (probably non-public, but at least not named with `internal`) headers;
3. Add `ifdef` protection for `DCHECK` definition like we did for `ASSIGN_OR_RAISE` (https://github.com/apache/arrow/blob/03726178494c8978bf48b9bab15ed9676e7c9196/cpp/src/arrow/result_internal.h#L20-L22) and `RETURN_NOT_OK` (https://github.com/apache/arrow/blob/03726178494c8978bf48b9bab15ed9676e7c9196/cpp/src/arrow/status.h#L80-L82), to not mess up user code that directly includes `arrow/util/logging.h`.
4. Add comments as guideline.

### Are these changes tested?

No test needed.

### Are there any user-facing changes?

Probably not?

* GitHub Issue: #43209

### CODE REVIEW

File: cpp/build-support/lint_cpp_cli.py [hunk 1]
```diff
@@ -54,14 +55,12 @@ def lint_file(path):
         (lambda x: re.match(_RETURN_NOT_OK_REGEX, x),
          'Use ARROW_RETURN_NOT_OK in header files', _paths('''\
          arrow/status.h
-         test
-         arrow/util/hash.h
          arrow/python/util''')),
         (lambda x: re.match(_ASSIGN_OR_RAISE_REGEX, x),
-         'Use ARROW_ASSIGN_OR_RAISE in header files', _paths('''\
-         arrow/result_internal.h
-         test
```
Comment (zanmato1984) [inline cpp/build-support/lint_cpp_cli.py:63]: File whose path containing `test` and `internal` will be excluded universally in `lint_files`.

`hash.h` is already gone.

**Comment by zanmato1984** [comment]
A small cleanup. @pitrou would you like to take a look? Thanks.

**Review by ghost — APPROVED** [review]

**Comment by wgtmac** [comment]
@github-actions crossbow submit -g cpp

**Comment by zanmato1984** [comment]
@github-actions crossbow submit -g cpp

**Review by mapleFU — APPROVED** [review]
Thanks for your good catch!

File: cpp/src/arrow/acero/unmaterialized_table.h [hunk 2]
```diff
@@ -76,7 +76,7 @@ class UnmaterializedCompositeTable {
     if (Empty()) {
```
Comment (pitrou) [inline cpp/src/arrow/acero/unmaterialized_table.h:76]: We could try to make this file internal IMHO.

File: cpp/src/arrow/util/bit_stream_utils.h [hunk 3]
```diff
@@ -207,9 +207,9 @@ class BitReader {
 };
```
Comment (pitrou) [inline cpp/src/arrow/util/bit_stream_utils.h:207]: We could also try to make this file internal IMHO.

File: cpp/src/arrow/util/vector.h [hunk 4]
```diff
@@ -31,8 +31,8 @@ namespace internal {
```
Comment (pitrou) [inline cpp/src/arrow/util/vector.h:31]: We could also try to make this file internal IMHO.

File: cpp/src/arrow/acero/unmaterialized_table.h (hunk 2 above)
Comment (zanmato1984) [reply cpp/src/arrow/acero/unmaterialized_table.h:76]: Thank you. It always feels good to confirm something to be internal only. I'll rename them.

Comment (zanmato1984) [reply cpp/src/arrow/acero/unmaterialized_table.h]: Renamed.

File: cpp/src/arrow/util/bit_stream_utils.h (hunk 3 above)
Comment (zanmato1984) [reply cpp/src/arrow/util/bit_stream_utils.h]: Renamed this one and its transitive including `rle_encoding.h`.

File: cpp/src/arrow/util/vector.h (hunk 4 above)
Comment (zanmato1984) [reply cpp/src/arrow/util/vector.h:31]: I'm holding back this one, as it is included in two examples:
https://github.com/apache/arrow/blob/005125790d83ecf50ecc20026d97021d7eb19fe2/cpp/examples/arrow/join_example.cc#L40
https://github.com/apache/arrow/blob/005125790d83ecf50ecc20026d97021d7eb19fe2/cpp/examples/arrow/execution_plan_documentation_examples.cc#L48

And I tend to not argue its internality (and its several transitive headers') in this poor PR. @pitrou What's your opinion? Thanks.

Comment (pitrou) [reply cpp/src/arrow/util/vector.h:31]: Ok, let's keep it public then.

**Comment by pitrou** [comment]
@github-actions crossbow submit -g cpp

**Review by pitrou — APPROVED** [review]

**Comment by pitrou** [comment]
@zanmato1984 I think we should avoid `@`ing usernames in PR descriptions, because they end up in the final commit message and then generate notification spam when people update their forks.

**Comment by zanmato1984** [comment]
> @zanmato1984 I think we should avoid `@`ing usernames in PR descriptions, because they end up in the final commit message and then can generate notification spam when people update their forks.

Thanks for reminding. That's very thoughtful and I have updated the PR description. Appreciate that!

**Comment by pitrou** [comment]
Thanks a lot @zanmato1984 !

**Pull request merged**